src/Pate/Proof.hs

```diff
@@ -353,296 +378,162 @@ traverseProofApp f = \case
-    <*> pure a4
-    
-  ProofStatus a1 -> ProofStatus
-    <$> traverse (\(cex, cond) -> (,) <$> prfCounterExampleTrans f cex <*> prfConditionTrans f cond) a1
-  ProofDomain a1 a2 a3 a4 -> ProofDomain
-    <$> MM.traverseRegsWith (\_ (Const v) -> Const <$> prfPredTrans f v) a1
-    <*> transformMemDomain f a2
-    <*> transformMemDomain f a3
-    <*> prfContextTrans f a4
-
-
 -- | A 'ProofExpr' is an direct proof representation, where
 -- nodes hold completed sub-proofs.
-data ProofExpr prf tp where
-  ProofExpr :: { unApp :: ProofApp prf (ProofExpr prf) tp } -> ProofExpr prf tp
+data ProofExpr sym arch tp where
+  ProofExpr :: { unApp :: ProofApp sym arch (ProofExpr sym arch) tp } -> ProofExpr sym arch tp
 
 -- | Traverse all sub-expressions of a 'ProofExpr'
 traverseProofExpr ::
   Monad m =>
-  (forall tp'. ProofExpr prf tp' -> m (ProofExpr prf tp')) ->
-  ProofExpr prf tp ->
-  m (ProofExpr prf tp)
+  (forall tp'. ProofExpr sym arch tp' -> m (ProofExpr sym arch tp')) ->
+  ProofExpr sym arch tp ->
+  m (ProofExpr sym arch tp)
 traverseProofExpr f e =
   ProofExpr <$> (traverseProofApp (traverseProofExpr f) =<< (unApp <$> f e))
 
-transformProofExpr ::
-  Monad m =>
-  ProofTransformer m prf prf' ->
-  ProofExpr prf tp ->
-  m (ProofExpr prf' tp)
-transformProofExpr f (ProofExpr app) =
-  ProofExpr <$> (traverseProofApp (transformProofExpr f) =<< (transformProofApp f app))
-
 -- | Collect results from all sub-expressions
 collectProofExpr ::
-  forall w prf tp.
+  forall w sym arch tp.
   Monoid w =>
-  (forall tp'. ProofExpr prf tp' -> w) ->
-  ProofExpr prf tp ->
+  (forall tp'. ProofExpr sym arch tp' -> w) ->
+  ProofExpr sym arch tp ->
   w
 collectProofExpr f e_outer = runIdentity $ CMW.execWriterT (traverseProofExpr go e_outer)
   where
-    go :: ProofExpr prf tp' -> CMW.WriterT w Identity (ProofExpr prf tp')
+    go :: ProofExpr sym arch tp' -> CMW.WriterT w Identity (ProofExpr sym arch tp')
     go e = do
       CMW.tell (f e)
       return e
 
--- | A domain expression is actually independent of the 'app' parameter, so we can
--- project it to any one.
-appDomain ::
-  ProofExpr prf ProofDomainType ->
-  ProofApp prf app ProofDomainType
-appDomain (ProofExpr (ProofDomain regs stack glob domCtx)) = ProofDomain regs stack glob domCtx
-
-emptyDomain ::
-  forall prf m.
-  Monad m =>
-  IsProof prf =>
-  IsBoolLike prf m =>
-  ProofContext prf ->
-  m (ProofExpr prf ProofDomainType)
-emptyDomain domCtx = do
-  regs <- MM.mkRegStateM (\_ -> Const <$> proofPredFalse @prf)
-  stack <- emptyMemDomain
-  globs <- emptyMemDomain
-  return $ ProofExpr $ ProofDomain regs stack globs domCtx
-
-emptyMemDomain ::
-  forall prf m.
-  Monad m =>
-  IsBoolLike prf m =>
-  m (ProofMemoryDomain prf)
-emptyMemDomain = do
-  falsePred <- proofPredFalse @prf
-  return $ ProofMemoryDomain MapF.empty falsePred
-
-type family ProofScope prf :: DK.Type
+type family SymScope sym :: DK.Type
+type instance SymScope (W4B.ExprBuilder t fs scope) = t
```

Comment:
It is a bit hard to eyeball where this is used. On one hand it is kind of funny that `SymScope` doesn't extract the type parameter named `scope`.

If this is only used in a few places (e.g., `ProofNonce` is the only one I saw). If that really is the only use, could we avoid these potentially confusing types by just defining
```
data ProofNonce sym (tp :: ProofNodeType) where
  ProofNonce :: (sym ~ W4B.ExprBuilder t fs scope) => N.Nonce t -> ProofNonce sym tp
```

Reply:
If it is used widely and is important, fair enough...

Reply:
It's used in at least one other place, which is to supply the type parameter to the nonce generator in `Pate.Monad.Environment`. It's not obvious to me how to write that easily without the type family.